### PR TITLE
Scalar crypto - post code review improvments

### PR DESCRIPTION
**Pull request by ben-marshall**
Hi there

This PR contains updates from @jrtc27's review [here](https://github.com/riscv/sail-riscv/pull/99#pullrequestreview-786873720) which were missing from the original merge of the scalar cryptography extension code.

- [Commit 1](https://github.com/riscv/sail-riscv/commit/a527ad769c9acba9cba70261c3bd73a1d5a0069c) addresses the [inconsistency of `==` whitespace](https://github.com/riscv/sail-riscv/pull/99#discussion_r734560390)
- [Commit 2](https://github.com/riscv/sail-riscv/commit/c4ad3c16a943de9996224a2439b5488ce4fac359) addresses the [inconsistency of ` );`](https://github.com/riscv/sail-riscv/pull/99#discussion_r734563951) whitespace in the `sha*` instructions.
- [Commit 3](https://github.com/riscv/sail-riscv/commit/53611ff431bac23b438f30bbd214cff3c55c6bb6) tries to address [comments](https://github.com/riscv/sail-riscv/pull/99#discussion_r734584700) about the clarity of `aesks1i` by adding code comments, and ensuring that the `aes_decode_rcon` function only accepts valid inputs. It also renames some arguments to the `aesks1i` execute and encdec clauses to match the specification.

About [this question](https://github.com/riscv/sail-riscv/pull/99#discussion_r734562267) asking "please confirm every single change in [commit](https://github.com/riscv/sail-riscv/compare/33aaaa04eaeaa9d2ed69023735ad6ebb93f71775..b83f68b3ade2d5444fd7dfe952e441824bb19dd8) is what you intended?" - Yes, these functional changes were intended and should have been part of the initial PR. Specifically, the changes to when `aesks1i` and `aesks2` are enabled were deliberate, as they should be present for both the `Zkne` and `Zknd` extensions.

About the `aesks1i` comments [here](https://github.com/riscv/sail-riscv/pull/99#discussion_r734584700) - It's been interesting watching people react to these instructions. The more cryptography minded people have reacted quite positively to how they're specified because they map onto some common implementation patterns for the AES KeySchedule _fairly_ well. But, that's only apparent if you're very familiar with those patterns in the first place, which is precious few people. Others have initially been quite perplexed about them for exactly the reasons highlighted. If there's anything _more_ to be done about explaining these instructions better then I'm all ears. Whether this is best going into the Sail code or the specification document, I'm not sure - maybe both. In terms of being future-proof to changes in FIPS-197 parameters, I think we are fairly safe. AES is very stable at this point, and I don't think NIST or anyone in the cryptography community is expecting to see AES parameter changes.

I've deliberately kept the commits separate in the hope it makes the PR easier to map back onto the original review. If you'd like me to squash them down before merging, please let me know. Likewise if there's anything else which could be made clearer.

Cheers,
Ben

### CODE REVIEW

**Comment by jrtc27** [comment]
Do you have an opinion on:
> I'd probably rather that be specified in a more obvious manner, with aes_decode_rcon only accepting legal inputs instead of silently mapping reserved ones to 0, and AES64KS1I having a very clear special case for 0xA, rather than trying to shoe-horn it into the normal case.

and

> That is, I think this would be a lot easier to understand (and easier to know how to use; I suspect people who know about AES-256 but not Zkn will initially be confused like I was and think it's broken for AES-256 like I did, but maybe that's just me) as something more like:
> ```
> function clause execute (AES64KS1I(rnum, rs1, rd)) = {
>   assert(sizeof(xlen) == 64);
>   let prev     : bits(32) = X(rs1)[63..32];
>   let subwords : bits(32) = aes_subword_fwd(prev);
>   let result   : bits(32) = if (rnum == 0xA) then subwords
>                             else (subwords >>> 8) ^ aes_decode_rcon(rnum);
>   X(rd) = result @ result;
>   RETIRE_SUCCESS
> }
> ```
> I still don't like the duplication of state in rd given the low half is entirely unused (at which point, really, the instructions should just be producing the output in the low 32 bits and zeroes in the others), which I assume is just left over from AES-NI inspiration, but that is probably too big a change at this stage of the spec, unfortunately, so we have to live with that wart (alongside the 0xA sentinel precluding future extension).
> 
> I also took the liberty of renaming some things to what I believe are better names in that example (and rcon -> rnum to match the spec change that came out of one of our previous discussions).

I think that, combined with a brief single sentence note calling out the special 0xA case would go a long way to make things clear for those not intimately familiar with AES.

**Comment by ben-marshall** [comment]
I agree that adding a sentence to explicitly say the `0xA` case is particular for AES-256 is a helpful pointer for people.

In terms of restructuring the `aesks1i` code. I don't have a strong opinion, but I am happy being guided by your experience in trying to understand how the instruction works. If the code you provided is clearer to you (and we assume someone familiar with AES-256 would get it anyway) then that's the better way to write it. I'll make these changes in a commit shortly.

I'm sorry I missed this suggestion initially, Github has an annoying habit of collapsing review comments & threads.

File: model/riscv_types_kext.sail [hunk 1]
```diff
@@ -68,8 +68,14 @@ function aes_mixcolumn_inv(x) = {
   b3 @ b2 @ b1 @ b0 /* Return value */
 }
 
-val aes_decode_rcon : bits(4) -> bits(32)
+/* Turn a round number into a round constant for AES. Note that the
+   AES64KS1I instruction is defined such that the r argument is always
+   in the range 0x0..0xA. Values of rnum outside this range do not decode
+   to the AES64KS1I instruction. The 0xA => 0x0 case is used specifically
```
Comment (jrtc27) [inline model/riscv_types_kext.sail:74]: This special case should no longer be hit thanks to the revised AES64KS1I implementation, so you can have an assert that it's less than 0xA

**Review by martinberger — APPROVED** [review]
OK with me.

File: model/riscv_types_kext.sail (hunk 1 above)
Comment (ben-marshall) [reply model/riscv_types_kext.sail:74]: Ah right you are! Duly change in the most recent push. :+1:

File: model/riscv_types_kext.sail [hunk 2]
```diff
@@ -68,8 +68,15 @@ function aes_mixcolumn_inv(x) = {
   b3 @ b2 @ b1 @ b0 /* Return value */
 }
 
-val aes_decode_rcon : bits(4) -> bits(32)
+/* Turn a round number into a round constant for AES. Note that the
+   AES64KS1I instruction is defined such that the r argument is always
+   in the range 0x0..0xB. Values of rnum outside the range 0x0..0xA
+   do not decode to the AES64KS1I instruction. The 0xA => 0x0 case is used
```
Comment (jrtc27) [inline model/riscv_types_kext.sail:74]: > The 0xA => 0x0 case is used

There is no 0xA => 0x0 case, the function is just never called and no XOR is ever performed.

File: model/riscv_types_kext.sail [hunk 3]
```diff
@@ -68,8 +68,15 @@ function aes_mixcolumn_inv(x) = {
   b3 @ b2 @ b1 @ b0 /* Return value */
 }
 
-val aes_decode_rcon : bits(4) -> bits(32)
+/* Turn a round number into a round constant for AES. Note that the
+   AES64KS1I instruction is defined such that the r argument is always
+   in the range 0x0..0xB. Values of rnum outside the range 0x0..0xA
```
Comment (jrtc27) [inline model/riscv_types_kext.sail:73]: The first one should be 0xA not 0xB?

File: model/riscv_types_kext.sail (hunk 2 above)
Comment (jrtc27) [reply model/riscv_types_kext.sail:74]: (i.e. just drop the => 0 and I think the sentence reads fine?)

File: model/riscv_types_kext.sail (hunk 3 above)
Comment (ben-marshall) [reply model/riscv_types_kext.sail]: Doh. Right you are

File: model/riscv_types_kext.sail (hunk 2 above)
Comment (ben-marshall) [reply model/riscv_types_kext.sail]: Agreed, just fixed this now.

**Review by jrtc27 — APPROVED** [review]
Thanks, looks good to me now

**Comment by martinberger** [comment]
> Thanks, looks good to me now

Thanks!

**Pull request merged**